embassy-nrf/src/time_driver.rs

```diff
@@ -83,194 +79,407 @@ impl AlarmState {
+
+        fn next_period(&self) {
+            critical_section::with(|cs| {
+                let r = timer();
+                let period = self.period.fetch_add(1, Ordering::Relaxed) + 1;
+                let t = (period as u64) << 23;
+
+                for n in 0..ALARM_COUNT {
+                    let alarm = &self.alarms.borrow(cs)[n];
+                    let at = alarm.timestamp.get();
 
-        match id {
-            Ok(id) => Some(AlarmHandle::new(id)),
-            Err(_) => None,
+                    if at < t + 0xc00000 {
+                        // just enable it. `set_alarm` has already set the correct CC val.
+                        r.intenset.write(|w| unsafe { w.bits(compare_n(n)) });
+                    }
+                }
+            })
         }
-    }
 
-    fn set_alarm_callback(&self, alarm: AlarmHandle, callback: fn(*mut ()), ctx: *mut ()) {
-        critical_section::with(|cs| {
-            let alarm = self.get_alarm(cs, alarm);
+        fn get_alarm<'a>(&'a self, cs: CriticalSection<'a>, alarm: AlarmHandle) -> &'a AlarmState {
+            // safety: we're allowed to assume the AlarmState is created by us, and
+            // we never create one that's out of bounds.
+            unsafe { self.alarms.borrow(cs).get_unchecked(alarm.id() as usize) }
+        }
 
-            alarm.callback.set(callback as *const ());
-            alarm.ctx.set(ctx);
-        })
-    }
+        fn trigger_alarm(&self, n: usize, cs: CriticalSection) {
+            let r = timer();
+            r.intenclr.write(|w| unsafe { w.bits(compare_n(n)) });
 
-    fn set_alarm(&self, alarm: AlarmHandle, timestamp: u64) {
-        critical_section::with(|cs| {
-            let n = alarm.id() as _;
-            let alarm = self.get_alarm(cs, alarm);
-            alarm.timestamp.set(timestamp);
+            let alarm = &self.alarms.borrow(cs)[n];
+            alarm.timestamp.set(u64::MAX);
 
-            let t = self.now();
+            // Call after clearing alarm, so the callback can set another alarm.
 
-            // If alarm timestamp has passed, trigger it instantly.
-            if timestamp <= t {
-                self.trigger_alarm(n, cs);
-                return;
-            }
+            // safety:
+            // - we can ignore the possiblity of `f` being unset (null) because of the safety contract of `allocate_alarm`.
+            // - other than that we only store valid function pointers into alarm.callback
+            let f: fn(*mut ()) = unsafe { mem::transmute(alarm.callback.get()) };
+            f(alarm.ctx.get());
+        }
+    }
 
-            let r = rtc();
+    impl Driver for TimerDriver {
+        fn now(&self) -> u64 {
+            // `period` MUST be read before `counter`, see comment at the top for details.
+            let period = self.period.load(Ordering::Relaxed);
+            compiler_fence(Ordering::Acquire);
+            let r = timer();
+            r.tasks_capture[3].write(|w| unsafe { w.bits(1) });
```

Comment:
This will overwrite whatever's in CC3 which you're currently using for (half of the) overflow handling. It's super unfortunate that you can't read the counter value directly. WTF Nordic? but it is what it is. We have to sacrifice a CC channel to read the counter >_<

This means it can't support 3 alarms unlike RTC. It can max 2 or 1 depending on how you implement overflows